doc/service-proxy.md

```diff
@@ -59,30 +67,42 @@ NAT'ing (especially SNAT'ing) that would otherwise be needed.
 
 A service may accept connections to a variety of IPs and ports:
 
-  - the service's `.spec.clusterIPs`, on any of the
-    `.spec.ports[*].port`s
+  - the service's `.spec.clusterIPs`, on any port listed in
+    `.spec.ports[*].port`
 
-  - any of the service's `.spec.externalIPs`, on any of the 
-    `.spec.ports[*].port`s
+  - any of the service's `.spec.externalIPs`, on any port listed in
+    `.spec.ports[*].port`
 
-  - any of the service's `.status.loadBalancer.ingress[*].ip`s, on any
-    of the `.spec.ports[*].port`s
+  - any of the service's `.status.loadBalancer.ingress[*].ip` IPs, on
+    any port listed in `.spec.ports[*].port`
 
-  - any local IP, on any of the service's `.spec.ports[*].nodePort`s
+  - any local IP, on any port listed in `.spec.ports[*].nodePort`
 
       - (Actually, _by default_ kube-proxy accepts NodePort
         connections on any IP, but it also allows you to restrict
         NodePorts to a specific subset of node IPs by passing the
         `--nodeport-addresses` flag.)
 
-If a service has at least one endpoint that is "ready" (according to
-its `.endpoints[].conditions.ready`), then connections to any of the
-above destinations should be randomly DNAT'ed to one of the ready
-endpoint IPs (with the destination port changed to the corresponding
-`.spec.port[].targetPort` value, if that is set). If a service does
-not have any "ready" endpoints, then connections to any of the above
-destinations should be rejected (ie, actively refused with an ICMP
-error; not simply dropped).
+      - The iptables backend of kube-proxy even allows NodePort
+        connections on `127.0.0.1` (but not `::1`), though the ipvs
+        backend does not, and this behavior is considered deprecated
+        and can be disabled.
+
+If a service has at least one usable endpoint, then connections to any
+of the above destinations should be randomly DNAT'ed to one of the
+usable endpoint IPs (with the destination port changed to the
+corresponding `.spec.port[].targetPort` value, if that is set). If a
+service does not have any usable endpoints, then connections to any of
+the above destinations should be rejected (ie, actively refused with
```

Comment:
Should the ICMP error always be a [Type 3 - Destination Unreachable ](https://www.iana.org/assignments/icmp-parameters/icmp-parameters.xhtml#icmp-parameters-codes-3) packet? Is that specified somewhere?

Reply:
I think that is what iptables does by default with `-j REJECT` and we carrying over with that , maybe we should indicate that it should reject the packet and leave that open, you can also use TCP_RST

Reply:
Right for UDP I think we'd have to use ICMP but for TCP that makes sense

Reply:
no, but... none of the other ICMP types really imply "reject" so it's strongly implied.

(Hm... and looking at the [ICMP code to errno conversion table](https://github.com/torvalds/linux/blob/v6.1/net/ipv4/icmp.c#L113) it looks like if we returned `ICMP_PORT_UNREACH` instead of the default `ICMP_HOST_UNREACH` then the caller would get back "Connection refused" instead of "No route to host"... Maybe we should start doing that, and recommend that...)

Reply:
ICMP_PORT_UNREACH is correct, you connect to a service IP:port and the port is not reachable because doesn't have endpoints

Reply:
is the iptables default for `-j REJECT` , isn't it?